### PR TITLE
Add `uv tool upgrade` command

### LINKED ISSUE #5188: Add `uv tool upgrade` to upgrade a tool
e.g. given `uv tool upgrade <package name>` we'd upgrade the tool associated with the package to the latest version within the constraints. The user cannot specify new constraints here, they'd need to use `uv tool install` for that.

### PR DESCRIPTION
**Pull request by blueraft**
## Summary

Resolves #5188. Most of the changes involve creating a new function in `tool/common.rs` to contain the common functionality previously found in `tool/install.rs`.

## Test Plan

`cargo test`

```console
❯ ./target/debug/uv tool upgrade black
warning: `uv tool upgrade` is experimental and may change without warning.
Resolved 6 packages in 25ms
Uninstalled 1 package in 3ms
Installed 1 package in 19ms
 - black==23.1.0
 + black==24.4.2
Installed 2 executables: black, blackd
```

### CODE REVIEW

File: crates/uv-cli/src/lib.rs [hunk 1]
```diff
@@ -2231,6 +2233,27 @@ pub struct ToolUninstallArgs {
...
+    /// The name of the tool to upgrade.
+    #[arg(required = true)]
+    pub name: Option<PackageName>,
+
+    /// Upgrade all tools.
+    #[arg(long, conflicts_with("name"))]
+    pub all: bool,
+
+    #[command(flatten)]
+    pub installer: ResolverInstallerArgs,
```
Comment (zanieb) [inline crates/uv-cli/src/lib.rs:2248]: Does including this mean we support things like `--upgrade-package` here now? Is that what we want?

File: crates/uv/src/settings.rs [hunk 2]
```diff
@@ -299,6 +302,40 @@ impl ToolListSettings {
     }
 }
 
+/// The resolved settings to use for a `tool uninstall` invocation.
```
Comment (zanieb) [inline crates/uv/src/settings.rs:305]:
```suggestion
/// The resolved settings to use for a `tool upgrade` invocation.
```

File: crates/uv/tests/tool_upgrade.rs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+    ----- stderr -----
+    warning: `uv tool upgrade` is experimental and may change without warning.
+    Resolved [N] packages in [TIME]
+    Prepared [N] packages in [TIME]
+    Uninstalled [N] packages in [TIME]
+    Installed [N] packages in [TIME]
+     - black==23.1.0
+     + black==24.3.0
+    Installed 2 executables: black, blackd
```
Comment (zanieb) [inline crates/uv/tests/tool_upgrade.rs:59]: Should these say "Updated"?

File: crates/uv/src/commands/tool/upgrade.rs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+use uv_normalize::PackageName;
+use uv_requirements::RequirementsSpecification;
+use uv_tool::InstalledTools;
+use uv_warnings::warn_user_once;
+
+use crate::commands::{tool::common::install_executables, ExitStatus, SharedState};
+use crate::printer::Printer;
+
+/// Upgrade a tool.
+pub(crate) async fn upgrade(
```
Comment (zanieb) [inline crates/uv/src/commands/tool/upgrade.rs:23]: I'll need to come back and read this more closely, sorry for a partial review.

File: crates/uv-cli/src/lib.rs (hunk 1 above)
Comment (blueraft) [reply crates/uv-cli/src/lib.rs:2248]: Good point. The following approach works, and there are cases where people might want to upgrade tools to a specific version as well. Maybe we should support this, wdyt?

```cargo run -- tool upgrade poetry --upgrade-package poetry==1.5```

File: crates/uv/tests/tool_upgrade.rs (hunk 3 above)
Comment (T-256) [reply crates/uv/tests/tool_upgrade.rs:59]:
```
Installed [N] packages in [TIME]
 - black==23.1.0
 + black==24.3.0
Installed 3 executables:
 ^ black
 ^ blackd
 + added-exe
 - removed-exe
```
same message format as package (un)installations diffs.
`^` mark stands for _Updated_ executable.

Comment (T-256) [reply crates/uv/tests/tool_upgrade.rs:59]: > `^` mark stands for _Updated_ executable.


As I'm thinking now, could we have this _Updated_ versions in package installations? example:
```
Installed [N] packages in [TIME]
 ^ black 23.1.0 -> 24.3.0
```

Comment (blueraft) [reply crates/uv/tests/tool_upgrade.rs]: I've gone with `Updated` for now. Once a symbol is settled in #5213 for `updated`, we can update this. Wdyt?

**Comment by charliermarsh** [comment]
Should `uv tool upgrade` work without arguments? That might mean that we need to store things like the index URLs in the install receipt. Otherwise, upgrades for tools installed from non-PyPI indexes won't work, unless you provide the index URL to `uv tool upgrade`.

**Comment by zanieb** [comment]
Yeah we probably need to store index URLs in the receipt 🤔 does pipx?

**Comment by blueraft** [comment]
These are the metadata stored by pipx, arguments like `--index-url` end up in the pip_args array:
```json
{
    "injected_packages": {},
    "main_package": {
        "app_paths": [
            {
                "__Path__": "/Users/ahmedilyas/Library/Application Support/pipx/venvs/sympy/bin/isympy",
                "__type__": "Path"
            }
        ],
        "app_paths_of_dependencies": {},
        "apps": [
            "isympy"
        ],
        "apps_of_dependencies": [],
        "include_apps": true,
        "include_dependencies": false,
        "man_pages": [
            "man1/isympy.1"
        ],
        "man_pages_of_dependencies": [],
        "man_paths": [
            {
                "__Path__": "/Users/ahmedilyas/Library/Application Support/pipx/venvs/sympy/share/man/man1/isympy.1",
                "__type__": "Path"
            }
        ],
        "man_paths_of_dependencies": {},
        "package": "sympy",
        "package_or_url": "sympy",
        "package_version": "1.13.1",
        "pip_args": [],
        "suffix": ""
    },
    "pipx_metadata_version": "0.3",
    "python_version": "Python 3.12.2",
    "venv_args": []
}
```

**Comment by zanieb** [comment]
Just noting this is blocked by https://github.com/astral-sh/uv/issues/5443

**Comment by charliermarsh** [comment]
I'm gonna be playing around with this branch as I'm trying to solve the "index URL" problem. Just a heads up @blueraft.

**Comment by blueraft** [comment]
All good!

**Review by charliermarsh — APPROVED** [review]
Thank you! Gonna build on this in some subsequent PRs. Merging with the known limitation that it doesn't respect prior settings.

**Pull request merged**